keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,981 @@
+
+### Notes/Constraints/Caveats
+
+This feature does not expand the ability of the NodeIPAM controller to change
+the `Node.Spec.PodCIDRs` field. Once that field is set, either by the controller
+or a third party, it will be treated as immutable. This is particularly relevant
+in situtaitons where users start modifying or deleting the `ClusterCidrConfig`.
+Under no circumstances will the controller attempt to revoke the allocated
+CIDRs (more details on this are discussed below).
+
+### Risks and Mitigations
+
+-   Racing kube-controller-managers. If multiples of the controller are running
+    (as in a HA control plane), how do they coordinate?
+    -   The controllers will coordinate using the existing
+        kube-controller-manager leader election.
+
+## Design Details
+
+### New Resource
+
+```go
+type ClusterCIDRConfig struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  Spec ClusterCIDRConfigSpec
+  Status ClusterCIDRConfigStatus
+}
+
+type ClusterCIDRConfigSpec {
+    // An IP block in CIDR notation ("10.0.0.0/8", "fd12:3456:789a:1::/64")
+    // +required
+    CIDR string
+
+    // This defines which nodes the config is applicable to. A nil selector can
+    // be applied to any node.
+    // +optional
+    NodeSelector *v1.LabelSelector
+
+    // Netmask size (e.g. 25 -> "/25") to allocate to a node.
+    // Users would have to ensure that the kubelet doesn't try to schedule
+    // more pods than are supported by the node's netmask (i.e. the kubelet's
+    // --max-pods flag)
+    // +required
+    PerNodeMaskSize int
+}
+
+type ClusterCIDRConfigStatus {
+  Conditions []metav1.Conditions
+}
+
+type ClusterCIDRConfigStatusType string
+type ClusterCIDRConfigStatusReason string
+var (
+  // If the "active" condition is true, this ClusterCIDRConfig can be used by
+  // the controller to allocate PodCIDRs for matching nodes.
+  ClusterCIDRConfigActive ClusterCIDRConfigStatusType = "active"
+
+  // If the "terminating" condition is true, this ClusterCIDRConfig was deleted
+  // by a user and is being garbage collected. When all Nodes using PodCIDRs
+  // from this range are deleted, the ClusterCIDRConfig will also be deleted.
+  ClusterCIDRConfigTerminating ClusterCIDRConfigStatusType = "terminating"
+
+  // If set as the reason on a false ClusterCIDRConfigActive condition, the
+  // ClusterCIDRConfig no longer has any free IP blocks.
+  ClusterCIDRConfigExhausted ClusterCIDRConfigStatusReason = "cidr-exhausted"
+
+  // If set as the reason on a true ClusterCIDRConfigTerminating condition,
+  // the ClusterCIDRConfig was used to allocate a Node's PodCIDR.
+  ClusterCIDRConfigInUse ClusterCIDRConfigStatusReason "cidr-in-use"
+)
+
+```
+
+#### Expected Behavior
+
+-   Each node will be assigned up to one range from each `FamilyType`. In case
+    of multiple matching ranges, attempt to break ties with the following rules:
+    1.  Pick the `ClusterCIDRRange` whose `PerNodeMaskSize` is the fewest IPs.
```

Comment:
could this be more elaborate and pick first the ClusterCIDRRange with lowest maxAvailableCIDRs in the set?
e.g. `CIDR=x.x.x.x/16, PerNodeMaskSize=16` (max=1) over `CIDR=x.x.x.x/24, PerNodeMaskSize=26`(max=4)

With your current description the second ClusterCIDRRange would be choosen.